[usb_host_hid] Implement USB HID support for US/ASCII keyboards

zopieux opened the pull request:
# What does this implement/fix?

esp32 platforms s2, s3, p4 support USB **host** HID, meaning HID devices such as keyboards and mouses can be connected and esp-idf can enumerate and poll them.

This PR adds basic support for capturing keyboard key presses for a standard ASCII (US) layout. It supports the 'shift' modifier for capital letters.

I was not able to make this play nicely with the existing `usb_host` component, because it also takes control of the devices when they connect, which is incompatible with esp-usb's HID implementation.

The `check_class` tweak is necessary for weirdly-behaved devices which act as a keyboard but do not report being keyboards, such as USB badge readers.

This feature was successfully tested on an esp32-s3 devkit.

Naturally, this component plays nicely with a `key_collector`:

```yaml
usb_host_hid:
  keyboards:
    - id: kbd
      # ...

key_collector:
  - id: reader
    source_id: kbd
    end_keys: "\r"
```

Next steps (purposefully not included) could include:

* exposing more of the HID device's properties as (text) sensors; this is trivial
* extending the support to mouse events.

This is my first interaction with ESPHome as both a user and developer, please consider tuning your feedback accordingly.

<details>

<summary>Example log output</summary>

```
[D]	[usb_hid:086]	Device connected: 1234:5428, Manufacturer: 'snip', Product: 'snip', Serial Number: '010100535755...'
[I]	[usb_hid:108]	Registered device 1234:5428 connected
[D]	[usb_hid:138]	Device 1234:5428 is listening for keys
[V]	[usb_hid:185]	Device 1234:5428 pressed key: 73, 'I'
[V]	[usb_hid:185]	(...)
[V]	[usb_hid:185]	Device 1234:5428 pressed key: 1
[D]	[main:240]	key collector finished! 'I love esphome'
[I]	[usb_hid:154]	Device 1234:5428 disconnected
```

</details>

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Partially implements https://github.com/esphome/feature-requests/issues/3003 (the wording is quite vague)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- https://github.com/esphome/esphome-docs/pull/5134

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
usb_host_hid:
  # enable_hubs: true
  keyboards:
    - id: kbd
      # check_class: false
      vid: 0xabcd
      pid: 0x1234

key_collector:
  - id: kbd_reader
    source_id: kbd
    end_keys: "\r"
    end_key_required: true
    timeout: 1s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs)

zopieux commented:
Wow that's not a very smart bot, sorry for the mention spam Clyde.

clydebarrow commented on esphome/const.py:
```diff
@@ -315,6 +315,7 @@ class PlatformFramework(Enum):
 CONF_EFFECTS = "effects"
 CONF_ELSE = "else"
 CONF_ENABLE_BTM = "enable_btm"
+CONF_ENABLE_HUBS = "enable_hubs"
```
Rather than putting these in core code, move to `components/const/__init__.py` - constants not used in core don't really need to be there, and if there are no core changes the PR is easier to test as an external component.

zopieux replied:
Sure thing.

zopieux committed:
[ush_host_hid] Implement USB HID support for US/ASCII keyboards

esp32 platforms s2, s3, p4 support USB **host** HID, meaning HID devices
such as keyboards and mouses can be connected and esp-idf can enumerate
and poll them.

This commit adds basic support for capturing keyboard key presses for a
standard ASCII (US) layout. It supports the 'shift' modifier for capital
letters.

I was not able to make this play nicely with the existing `usb_host`
component, because it also takes control of the devices when they
connect, which is incompatible with esp-usb's HID implementation.

The `check_class` tweak is necessary for weirdly-behaved devices which
act as a keyboard but do not report being keyboards, such as USB badge
readers.

This feature was successfully tested on an esp32-s3 devkit.

Naturally, this component plays nicely with a `key_collector`:

    usb_host_hid:
      keyboards:
        - id: kbd
          # ...

    key_collector:
      - id: reader
        source_id: kbd
        end_keys: "\r"

clydebarrow commented:
The reality is that with the number of open PRs and the limited resources to review them, some inactive PRs will fall to the bottom of the heap and be ignored anyway so the stale bot at least prompts you to refresh it.

Another way to gain attention for the PR is to post in the Discord dev channel. Not all potential reviewers would even be aware that a particular PR has been opened, but they're all active on Discord.
